src/java.base/share/classes/com/sun/crypto/provider/PBES2Parameters.java

```diff
@@ -344,27 +304,8 @@ protected byte[] engineGetEncoded() throws IOException {
         DerOutputStream out = new DerOutputStream();
 
         DerOutputStream pBES2_params = new DerOutputStream();
-
-        DerOutputStream keyDerivationFunc = new DerOutputStream();
-        keyDerivationFunc.putOID(pkcs5PBKDF2_OID);
-
-        DerOutputStream pBKDF2_params = new DerOutputStream();
-        pBKDF2_params.putOctetString(salt); // choice: 'specified OCTET STRING'
-        pBKDF2_params.putInteger(iCount);
-
-        if (keysize > 0) {
-            pBKDF2_params.putInteger(keysize / 8); // derived key length (in octets)
-        }
-
-        DerOutputStream prf = new DerOutputStream();
-        // algorithm is id-hmacWith<MD>
-        prf.putOID(kdfAlgo_OID);
-        // parameters is 'NULL'
-        prf.putNull();
-        pBKDF2_params.write(DerValue.tag_Sequence, prf);
-
-        keyDerivationFunc.write(DerValue.tag_Sequence, pBKDF2_params);
-        pBES2_params.write(DerValue.tag_Sequence, keyDerivationFunc);
+        pBES2_params.write(DerValue.tag_Sequence,
+                PBKDF2Parameters.encode(salt, iCount, keysize, kdfAlgo_OID));
```

Comment:
`keySize` here is number of bits, but the `PBKDF2Parameters.encode` requires number of bytes. This means a newly created PKCS12 keystore is invalid. When Java loads a PKCS12 keystore, this field is read but not validated. On the other hand, openssl validates it and will report an error.

Suggestion: either always use the same style in all fields and method arguments, or name them precisely.

Reply:
fixed